File: cleaningFunctions/set_merge.py
import os

import pandas as pd

CANONICAL_COLS = [
    "Tag",
    "Point number",
    "datetime",
    "type",
    "latitude",
    "longitude",
    "prev_lat",
    "prev_long",
    "colony_lat",
    "colony_long",
    "correct_step_distance",
]


def _series_to_iso_date_strings(s: pd.Series) -> pd.Series:
    """Parse datetimes (mixed ISO precision etc.), output YYYY-MM-DD strings."""
    ts = pd.to_datetime(s, utc=True, errors="coerce", format="mixed")
    if isinstance(ts.dtype, pd.DatetimeTZDtype):
        ts = ts.dt.tz_convert("UTC").dt.tz_localize(None)
    midnight = ts.dt.normalize()
    out = midnight.dt.strftime("%Y-%m-%d")
    return out.where(midnight.notna(), pd.NA)


def standardize_left(df: pd.DataFrame) -> pd.DataFrame:
    out = df.loc[:, [c for c in CANONICAL_COLS if c in df.columns]].copy()
    missing = set(CANONICAL_COLS) - set(out.columns)
    if missing:
        raise ValueError(f"left frame missing columns: {sorted(missing)}")
    out = out.reindex(columns=CANONICAL_COLS)
    out["datetime"] = _series_to_iso_date_strings(out["datetime"])
    return out


def standardize_og7(df: pd.DataFrame) -> pd.DataFrame:
    rename_map = {
        "BirdID": "Tag",
        "Data point #": "Point number",
        "Fix type": "type",
        "compensatedlat": "latitude",
        "long": "longitude",
    }
    missing_src = set(rename_map) - set(df.columns)
    if missing_src:
        raise ValueError(f"og7 frame missing columns: {sorted(missing_src)}")
    out = df.rename(columns=rename_map)
    required_before_datetime = [c for c in CANONICAL_COLS if c != "datetime"]
    for c in required_before_datetime:
        if c not in out.columns:
            raise ValueError(f"og7 frame missing column after rename: {c}")
    if "date" not in df.columns or "midvalue" not in df.columns:
        raise ValueError("og7 frame needs date and midvalue to build datetime")

    date_part = out["date"].astype(str).str.strip()
    time_part = out["midvalue"].astype(str).str.strip()
    combined = date_part + " " + time_part
    parsed = pd.to_datetime(combined, dayfirst=True, errors="coerce")
    out = out.reindex(columns=CANONICAL_COLS)
    out["datetime"] = _series_to_iso_date_strings(parsed)
    return out


def merge_chick_step_datasets(
    left_path: str,
    right_path: str,
) -> pd.DataFrame:
    left_df = pd.read_csv(left_path)
    right_df = pd.read_csv(right_path)
    left_std = standardize_left(left_df)
    right_std = standardize_og7(right_df)
    combined = pd.concat([left_std, right_std], ignore_index=True)
    combined["Tag"] = pd.to_numeric(combined["Tag"], errors="coerce").astype("Int64")
    combined["Point number"] = pd.to_numeric(combined["Point number"], errors="coerce").astype(
        "Int64"
    )
    return combined


def _default_paths() -> tuple[str, str]:
    root = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
    left = os.path.join(root, "original-datasets", "2013-2024correct_step_size.csv")
    right = os.path.join(root, "original-datasets", "og7chicks_nofilter_correctstepdistance.csv")
    return left, right


if __name__ == "__main__":
    left_p, right_p = _default_paths()
    combined = merge_chick_step_datasets(left_p, right_p)
    out_root = os.path.abspath(os.path.join(os.path.dirname(__file__), ".."))
    out_xlsx = os.path.join(out_root, "mergedchickdata(o7&2013-2014).xlsx")
    combined.to_excel(out_xlsx, index=False)

File: cleaningFunctionsR/set_merge.R
# Merge chick step datasets (matches cleaningFunctionsPython/set_merge.py).
# Run from repo root: Rscript cleaningFunctionR/set_merge.R
#
# Optional: install.packages("writexl") for Excel output.

CANONICAL_COLS <- c(
  "Tag",
  "Point number",
  "datetime",
  "type",
  "latitude",
  "longitude",
  "prev_lat",
  "prev_long",
  "colony_lat",
  "colony_long",
  "correct_step_distance"
)

repo_root <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  f <- grep("^--file=", args, value = TRUE)
  if (length(f)) {
    dirname(dirname(normalizePath(sub("^--file=", "", f))))
  } else {
    normalizePath(file.path(getwd()))
  }
}

#' Parse ISO8601 datetime strings; supports fractional seconds (mixed precision).
parse_iso_to_date_str <- function(x) {
  x <- trimws(as.character(x))
  out <- rep(NA_character_, length(x))
  opts <- options(digits.secs = 6)
  on.exit(options(opts))
  for (i in seq_along(x)) {
    xi <- x[i]
    if (is.na(xi) || !nzchar(xi)) next
    pt <- NA
    for (fmt in c("%Y-%m-%dT%H:%M:%OSZ", "%Y-%m-%dT%H:%M:%SZ")) {
      tmp <- tryCatch(
        as.POSIXct(xi, tz = "UTC", format = fmt),
        warning = function(w) NA,
        error = function(e) NA
      )
      if (!is.na(tmp)) {
        pt <- tmp
        break
      }
    }
    if (!is.na(pt)) {
      out[i] <- format(as.Date(pt, tz = "UTC"), "%Y-%m-%d")
    }
  }
  out
}

#' Parse og7 date + midvalue (day-first, UK-style dates).
parse_og7_datetime_to_date_str <- function(date_part, time_part) {
  d <- trimws(as.character(date_part))
  t <- trimws(as.character(time_part))
  n <- length(d)
  out <- rep(NA_character_, n)
  opts <- options(digits.secs = 6)
  on.exit(options(opts))
  for (i in seq_len(n)) {
    if (isTRUE(is.na(d[i]) && is.na(t[i]))) next
    combined <- paste(d[i], t[i])
    pt <- NA
    for (fmt in c("%d/%m/%Y %H:%M:%OS", "%d/%m/%Y %H:%M", "%d-%m-%Y %H:%M", "%Y-%m-%d %H:%M")) {
      tmp <- tryCatch(
        as.POSIXct(combined, tz = "UTC", format = fmt),
        warning = function(w) NA,
        error = function(e) NA
      )
      if (length(tmp) == 1L && !is.na(tmp)) {
        pt <- tmp
        break
      }
    }
    if (length(pt) == 1L && !is.na(pt)) {
      out[i] <- format(as.Date(pt, tz = "UTC"), "%Y-%m-%d")
    }
  }
  out
}

standardize_left <- function(df) {
  miss <- setdiff(CANONICAL_COLS, names(df))
  if (length(miss)) stop("left frame missing columns: ", paste(sort(miss), collapse = ", "))
  out <- df[, CANONICAL_COLS, drop = FALSE]
  out$datetime <- parse_iso_to_date_str(out$datetime)
  out
}

standardize_og7 <- function(df) {
  rename_map <- c(
    "BirdID" = "Tag",
    "Data point #" = "Point number",
    "Fix type" = "type",
    "compensatedlat" = "latitude",
    "long" = "longitude"
  )
  miss <- setdiff(names(rename_map), names(df))
  if (length(miss)) stop("og7 frame missing columns: ", paste(sort(miss), collapse = ", "))
  if (!all(c("date", "midvalue") %in% names(df))) {
    stop("og7 frame needs date and midvalue to build datetime")
  }

  out <- df
  for (nm in names(rename_map)) {
    names(out)[names(out) == nm] <- rename_map[[nm]]
  }

  need <- setdiff(CANONICAL_COLS[CANONICAL_COLS != "datetime"], names(out))
  if (length(need)) stop("og7 frame missing column after rename: ", paste(need, collapse = ", "))

  dt_str <- parse_og7_datetime_to_date_str(out$date, out$midvalue)
  out <- out[, CANONICAL_COLS[CANONICAL_COLS != "datetime"], drop = FALSE]
  out$datetime <- dt_str
  out <- out[, CANONICAL_COLS, drop = FALSE]
  out
}

merge_chick_step_datasets <- function(left_path, right_path) {
  left_df <- utils::read.csv(left_path, stringsAsFactors = FALSE, check.names = FALSE)
  right_df <- utils::read.csv(right_path, stringsAsFactors = FALSE, check.names = FALSE)

  left_std <- standardize_left(left_df)
  right_std <- standardize_og7(right_df)

  combined <- rbind(left_std, right_std)

  combined$Tag <- suppressWarnings(as.integer(combined$Tag))
  combined$`Point number` <- suppressWarnings(as.integer(combined$`Point number`))

  lon <- suppressWarnings(as.numeric(combined$longitude))
  combined$calculation_long <- ((lon %% 360) + 360) %% 360

  combined
}

default_paths <- function(root) {
  left <- file.path(root, "original-datasets", "2013-2024correct_step_size.csv")
  right <- file.path(root, "original-datasets", "og7chicks_nofilter_correctstepdistance.csv")
  list(left = left, right = right)
}

launched_as_script <- function() {
  args <- commandArgs(trailingOnly = FALSE)
  any(grepl("^--file=", args))
}

if (launched_as_script()) {
  root <- repo_root()
  p <- default_paths(root)
  combined <- merge_chick_step_datasets(p$left, p$right)

  out_csv <- file.path(root, "mergedchickdata(o7&2013-2014).csv")
  utils::write.csv(combined, out_csv, row.names = FALSE, na = "")

  if (requireNamespace("writexl", quietly = TRUE)) {
    writexl::write_xlsx(combined, file.path(root, "mergedchickdata(o7&2013-2014).xlsx"))
  } else {
    message("Install 'writexl' for mergedchickdata Excel output (CSV written).")
  }
}
